### PR TITLE
Fix Segfault on Windows 64 bits

### PR DESCRIPTION
**Pull request by mlasson**
The segfault happens when you use a wildcard '*' or '?'
in the argument of the caml program compiled on Windows
64bits.

The handle was corrupted because it is cast from
intptr_t (64bits) to int (32bits).

### CODE REVIEW

**Review by shindere — APPROVED** [review]

**Comment by shindere** [comment]
Thanks! I think that one is worth a Changes entry

**Comment by nojb** [comment]
This PR makes me wonder if we should be concerned with https://github.com/mlasson/ocaml/blob/54558fdc1f042bf48b0302e30a9bf91ab3749634/byterun/win32.c#L61.

**Comment by mlasson** [comment]
@nojb: ok, also it seems that my change break the compilation on the old compiler, I should probably add the #ifdef too

@shindere ok ! I will add a non-regression test and a change entry tomorrow

**Comment by shindere** [comment]
Marc Lasson (2018/02/20 18:44 +0000):
> @shindere ok ! I will add a non-regression test and a change entry
> tomorrow

Great, thanks!

Could you please use ocamltest to write that test?
If you have any problem doing that I will be pleased to help.

File: byterun/win32.c
Comment (nojb) [inline byterun/win32.c]: This is wrong; `long` is 32-bit on Windows, see https://msdn.microsoft.com/en-us/library/3b2e7499.aspx.

File: byterun/win32.c
Comment (nojb) [inline byterun/win32.c]: Same problem as above.

**Review by nojb — CHANGES_REQUESTED** [review]

**Comment by nojb** [comment]
More generally, are we sure that the fallback for when `intptr_t` is not defined is still necessary? Can this be the case for the supported versions of Windows? @dra27?

**Comment by dra27** [comment]
@nojb - the fallback is for old *compilers*, rather than old versions of Windows. It's becoming less likely to be hit, but I think you'd be surprised how many MSVC users are still tied to old versions for annoying reasons (I, for one, still have to use VS2003 for a project - though not an OCaml project)

**Comment by mlasson** [comment]
@dra27 : do you think it makes sense to keep my commit 1b35d7b48b2e425bb832781262eadb3892f6bec7 ?

**Comment by alainfrisch** [comment]
Assigning to @dra27 to make a decision about the need to support versions of msvc without intptr_t.

**Comment by alainfrisch** [comment]
Out of curiosity, which type was returned by _wfindfirst on older msvc without intptr_t?

File: byterun/win32.c
Comment (xavierleroy) [inline byterun/win32.c]: Just use `uintnat` and `intnat` defined in `caml/config.h`, these integer types are guaranteed to have the same size as pointer types:
```
typedef uintnat uintptr_t;
```

File: byterun/win32.c
Comment (xavierleroy) [inline byterun/win32.c]: Likewise, replace this `#ifdef _WIN64`...`#endif` by
```
typedef intnat intptr_t;
```

**Review by xavierleroy — COMMENTED** [review]
The fallback definitions of `intptr_t` and  `uintptr_t` can be  simplified.

**Review by xavierleroy — APPROVED** [review]
Looks good to me.  @dra27 @alainfrisch : would you like to have another look at this code, or can we merge as is?

**Review by nojb — APPROVED** [review]

File: byterun/win32.c
Comment (alainfrisch) [inline byterun/win32.c]: We still have the following in channels.c:
````c
#if defined(_MSC_VER) && !defined(_INTPTR_T_DEFINED)
typedef int intptr_t;
#define _INTPTR_T_DEFINED
#endif
....
...
int win_CRT_fd_of_filedescr(value handle)
{
  if (CRT_fd_val(handle) != NO_CRT_FD) {
    return CRT_fd_val(handle);
  } else {
    int fd = _open_osfhandle((intptr_t) Handle_val(handle), O_BINARY);
    if (fd == -1) uerror("channel_of_descr", Nothing);
    CRT_fd_val(handle) = fd;
    return fd;
  }
}
````

Wouldn't that cause a similar problem, on 64-bit and old enough versions of MSVC?  (Or perhaps all 64-bit versions do define intptr_t?)

Comment (nojb) [reply byterun/win32.c]: The `typedef` is wrong, but in principle it is safe: file `HANDLE`s in Win 64 are still 32-bits. See https://msdn.microsoft.com/en-us/library/windows/desktop/aa384203%28v=vs.85%29.aspx.

Comment (alainfrisch) [reply byterun/win32.c]: Ok.  Perhaps worth fixing it, but can be done independently of this PR (unless we prefer to add the missing typedef in a central place).

Comment (mlasson) [reply byterun/win32.c]: What about adding after the definition intnat and uintnat in config.h the following piece of code:
````
#ifndef _UINTPTR_T_DEFINED
typedef uintnat uintptr_t;
#define _UINTPTR_T_DEFINED
#endif
#ifndef _INTPTR_T_DEFINED
typedef intnat intptr_t;
#define _INTPTR_T_DEFINED
#endif
````
and remove all other typedef of intptr_t in the codebase ?

Comment (mlasson) [reply byterun/win32.c]: Or could simply use "intnat" and "uintnat" in place of intptr_t ?
(The only drawbacks I can think of is the risk to forget to replace it in future code).

Comment (xavierleroy) [reply byterun/win32.c]: > Or could simply use "intnat" and "uintnat" in place of intptr_t ?

That might be more prudent than defining `{u,}intptr_t` yourself, with the risk that some system header file doesn't understand what you did and tries to define it again.

On the other hand, if all recent enough versions of MSVC define those standard types, maybe we should just give up on supporting  the older versions.

Comment (alainfrisch) [reply byterun/win32.c]: I fully agree.  So, we need to decide between:

  - Use uintnat/intnat (unconditionally)
  - Use intptr_t/uintptr_t, which means no longer supporting old versions of MSVC

My understanding is that versions of MSVC that don't support intptr_t are also versions that don't come with a 64-bit variant, meaning they are really old.  Support could be restored purely in the build system by passing something like `/Dintptr_t=int` to cl.exe, if really needed.

@dra27 : you would be the (likely only) one to care for such old versions, so your input would be much appreciated!

**Comment by dra27** [comment]
Sorry - I should have checked this more closely more quickly, as there's been far too much thought invested in maintaining VC6 support here. Regarding `channels.c`, the typedef was intentionally simplified because no 64-bit headers are missing the definition, so it will always be 32-bit.

It's highly unlikely anyone's managed to write code which typedef's `uintptr_t` and `intptr_t` without also defining the guards and not noticed their error - they're in crtdefs.h which is included by *everything*. However, in principle I agree with @xavierleroy's caution.

I personally, even in C, prefer types to be being cast to the correct type, but I don't have a rational objection to using `intnat` or `uintnat` instead of `intptr_t` and `uintptr_t`, as long as it remains internal to the library/runtime.

**Comment by mshinwell** [comment]
@dra27 @alainfrisch Could you two come to a consensus as to how to proceed?

**Comment by alainfrisch** [comment]
As said, I **fully** agree with @xavierleroy 's proposal and @dra27 agrees **in principle**.  Let's call it a consensus!

**Comment by alainfrisch** [comment]
I think the consensus converges towards using uintnat/intnat (unconditionally) on each call site.

**Comment by nojb** [comment]
@mlasson could you rebase this PR, replacing all uses of `(u)intptr_t` with `(u)intnat` and remove all corresponding `#define`'s so we can merge this? Thanks!

**Comment by mlasson** [comment]
Yes, I will do that today !

**Comment by mlasson** [comment]
Ok. It fails to compile on 32bits (it does compile on 64 bits). 

The type equality: "uintptr_t = uintnat" is wrong on msvc 32bits... because uinptr_t expand to "unsigned int" (as define in vs headers) and uintnat to "unsigned long" (as defined in config.h). Note that both are 32 bits unsigned integer but they are not "equal" wrt. typechecking. 

So what about putting the polyfill macro (#if defined(_MSC_VER) && !defined(_UINTPTR_T_DEFINED) ...) in m.h (the one for win32 only), use (u)intptr_t when we need it ?

**Comment by alainfrisch** [comment]
> So what about putting the polyfill macro (#if defined(_MSC_VER) && !defined(_UINTPTR_T_DEFINED) ...) in m.h (the one for win32 only), use (u)intptr_t when we need it ?

This would work only if m.h included is included after the standard header that normally defines those types, no?

I'm leaning towards just using (u)intptr_t, which means that users who need to compile with very old versions of MSVC will need to do something (patch the code, or pass -Dintptr_t=...  to the C compiler).

At some point, we will want to require newer versions of MSVC anyway to benefit from better C99 support.

@dra27 : do you agree with this approach?

**Comment by mlasson** [comment]
I implemented @alainfrisch's suggestion. I putted the "polyfill" in m-nt.h / s-nt.h (I also added "HAS_STDINT_H" for MSC_VER >= 1600 when it was introduced).

File: config/m-nt.h
Comment (nojb) [inline config/m-nt.h]: `uintptr_t`

File: config/m-nt.h
Comment (nojb) [inline config/m-nt.h]: idem

File: config/m-nt.h
Comment (nojb) [inline config/m-nt.h]: I don't understand:
- `m.h` is included before any other header (see `caml/config.h`), so `_UINTPTR_T_DEFINED` will not be defined here even if they are defined later (in `stdint.h` I guess).
- Also, by defining `HAS_STDINT_H`, you are saying it is OK to include `stdint.h`, but you are defining it precisely in the case of old compilers, where it supposedly does not exist, agree?
- Why guard both against old compilers *and* undefined `intptr_t`? It seems to me that it would be enough to guard against one of those conditions...

File: config/s-nt.h
Comment (nojb) [inline config/s-nt.h]: Is this change related to this PR?

**Review by nojb — COMMENTED** [review]
Have some questions about the new version of the patch; I am not sure will work as intended in the present form.

File: config/s-nt.h
Comment (dra27) [inline config/s-nt.h]: This can just be `#if defined(__MINGW32__) || _MSC_VER >= 1600` - even in MS world, undefined macros are expanded to `0` in comparisons.

Comment (mlasson) [reply config/s-nt.h]: The idea is : either we use stdint.h, or we do the typedefs ourselves.

File: config/m-nt.h
Comment (dra27) [inline config/m-nt.h]: This guard is unnecessary (and it would be `1200`, not `1600`: Visual Studio 2010 added `stdint.h`, but these `typedef`s existed since Visual Studio .NET 2002 in, amongst others, `stdarg.h` and `stddef.h`)

**Review by dra27 — CHANGES_REQUESTED** [review]
Looks good to me apart from a minor nit and an agreement with @nojb's comment about removing the double-guard in `m-nt.h` (if the guard is to remain, then it should be academically correct - `1200`, not `1600`).

**Comment by dra27** [comment]
@alainfrisch - to answer your questions (finally!). We do include OS headers before `caml/config.h` - apart anything else, some of the headers depend upon it, I think. For your interest in history, `_wfindfirst` returned `long` in Visual C++ 6 (i.e. before `intptr_t` was added).

**Comment by nojb** [comment]
@mlasson could you address @dra27's comments, rebase to remove the useless intermediate commits and add @dra27, @alainfrisch and myself to the list of reviewers so that we can move to merge this? Thanks!

**Comment by dra27** [comment]
Just to throw one final spanner into these works. Just to be clear, the only Microsoft C Compiler which doesn't have definitions anywhere for `uintptr_t` and `intptr_t` and which is still just capable of building OCaml (or at least was in 2015 when I had too much time on my hands while performing some Messiahs...) is Visual C++ 6.0 (1998). Visual Studio .NET 2002 onwards have these - they're just in `stddef.h` prior to Visual Studio 2010. I would have no problem with altering the logic to define `HAS_STDINT_H` for Visual Studio 2010 and later (as this PR already does) but instead `#include <stddef.h>` for VS 2008 and earlier. i.e. in `config.h` have:

```C
#ifdef HAS_STDINT_H
#include <stdint.h>
#elif defined(_MSC_VER)
#include <stddef.h>
#endif
```
All the stuff in `m-nt.h` could then be dropped...

**Comment by mlasson** [comment]
Thanks @dra27 I like the last solution. 

The only thing that I would prefer to avoid is to pollute "config.h" with a _MSC_VER macro. 

What do you (or anyone else) think about one of these solutions: 
  0) include stddef.h in m-nt.h if version < 1200
  1) always include stddef.h in m-nt.h
  2) define "HAS_STDDEF_H" in s-nt.h and include stddef.h when present, 
  3) always include stddef.h in config.h
?

**Comment by dra27** [comment]
There are gcc-specific things in `config.h` already, so having an MS-specific one is not pollution :wink:

However, as `stddef.h` is a standard header, I can't envisage any harm in always including it in `config.h` (option 3) - could @xavierleroy (or another dev) quickly sign off on that too, and I think we're then good to go?

**Comment by damiendoligez** [comment]
> could @xavierleroy (or another dev) quickly sign off on that too, and I think we're then good to go?

If `stddef.h` is both part of C99 and available on MSVC, then yes let's include it.

**Comment by nojb** [comment]
> If stddef.h is both part of C99 and available on MSVC, then yes let's include it.

Yes, it is; so if @dra27 provides his formal approval, I will merge by the end of the day unless someone speaks up.

File: config/s-nt.h
Comment (dra27) [inline config/s-nt.h]: This **should** be `1600`.

`stdint.h` is only available from VS 2010 - the `1200` was to do with the `_UINTPTR_T_DEFINED` guards (since for `_MSC_VER < 1200`, `stddef.h` doesn't include the `intptr_t` types)

Comment (mlasson) [reply config/s-nt.h]: Ok, thanks ! It seems we are converging.

**Review by dra27 — APPROVED** [review]

**Comment by dra27** [comment]
Great, thank you - for the sake of completeness, I'll allow the CI to complete before merging.

**Comment by dra27** [comment]
Great, thank you! For completeness' sake, I'll allow the CI to complete before merging...

**Comment by dra27** [comment]
Cherry-picked to 4.07 in 33f576b9910a87df4e04a2af53a999358fbdf42f

**Pull request merged**
